pkg/reconciler/addressableservice/addressableservice.go

```diff
@@ -18,115 +18,49 @@ package addressableservice
 
 import (
 	"context"
-	"reflect"
 
-	"go.uber.org/zap"
-	corev1 "k8s.io/api/core/v1"
-	"k8s.io/apimachinery/pkg/api/equality"
 	apierrs "k8s.io/apimachinery/pkg/api/errors"
 	corev1listers "k8s.io/client-go/listers/core/v1"
-	"k8s.io/client-go/tools/cache"
-	"k8s.io/client-go/tools/record"
 	"knative.dev/pkg/apis"
 	duckv1beta1 "knative.dev/pkg/apis/duck/v1beta1"
-	"knative.dev/pkg/controller"
 	"knative.dev/pkg/logging"
 	"knative.dev/pkg/network"
 	"knative.dev/pkg/tracker"
-	"knative.dev/sample-controller/pkg/apis/samples/v1alpha1"
-	clientset "knative.dev/sample-controller/pkg/client/clientset/versioned"
-	listers "knative.dev/sample-controller/pkg/client/listers/samples/v1alpha1"
+	v1alpha1 "knative.dev/sample-controller/pkg/apis/samples/v1alpha1"
 )
 
 // Reconciler implements controller.Reconciler for AddressableService resources.
 type Reconciler struct {
-	// Client is used to write back status updates.
-	Client clientset.Interface
+	Core
 
 	// Listers index properties about resources
-	Lister        listers.AddressableServiceLister
 	ServiceLister corev1listers.ServiceLister
-
-	// The tracker builds an index of what resources are watching other
-	// resources so that we can immediately react to changes to changes in
-	// tracked resources.
-	Tracker tracker.Interface
-
-	// Recorder is an event recorder for recording Event resources to the
-	// Kubernetes API.
-	Recorder record.EventRecorder
```

Comment:
How should the user send events? Or put another way, why get rid of this?

Reply:
Core is inlined into the typed reconciler, they just do normal `r.Recorder(...` calls.